lib/SILOptimizer/Utils/OwnershipOptUtils.cpp

```diff
@@ -99,70 +99,88 @@ insertOwnedBaseValueAlongBranchEdge(BranchInst *bi, SILValue innerCopy,
   return phiArg;
 }
 
-static void getAllNonTrivialUsePointsOfBorrowedValue(
-    SILValue value, SmallVectorImpl<Operand *> &usePoints,
-    SmallVectorImpl<BorrowingOperand> &reborrowPoints) {
+static bool findTransitiveBorrowedUses(
+  SILValue value, SmallVectorImpl<Operand *> &usePoints,
+  SmallVectorImpl<BorrowingOperand> &reborrowPoints) {
   assert(value.getOwnershipKind() == OwnershipKind::Guaranteed);
 
   unsigned firstOffset = usePoints.size();
-  llvm::copy(value->getUses(), std::back_inserter(usePoints));
-
-  if (usePoints.size() == firstOffset)
-    return;
+  for (Operand *use : value->getUses()) {
+    if (use->getOperandOwnership() != OperandOwnership::NonUse)
+      usePoints.push_back(use);
+  }
 
   // NOTE: Use points resizes in this loop so usePoints.size() may be
   // different every time.
   for (unsigned i = firstOffset; i < usePoints.size(); ++i) {
-    if (auto fOperand = ForwardingOperand::get(usePoints[i])) {
-      fOperand.visitForwardedValues([&](SILValue transitiveValue) {
-        // Do not include transitive uses with 'none' ownership
-        if (transitiveValue.getOwnershipKind() == OwnershipKind::None)
-          return true;
-        for (auto *transitiveUse : transitiveValue->getUses())
-          usePoints.push_back(transitiveUse);
-        return true;
-      });
-      continue;
-    }
-
-    if (auto borrowingOp = BorrowingOperand::get(usePoints[i])) {
-      // If we have a reborrow, we have no further work to do, our reborrow is
-      // already a use and we will handle the reborrow separately.
-      if (borrowingOp.isReborrow())
-        continue;
+    Operand *use = usePoints[i];
+    switch (use->getOperandOwnership()) {
+    case OperandOwnership::NonUse:
+    case OperandOwnership::TrivialUse:
+    case OperandOwnership::ForwardingConsume:
+    case OperandOwnership::DestroyingConsume:
+      llvm_unreachable("this operand cannot handle a guaranteed use");
+
+    case OperandOwnership::ForwardingUnowned:
+    case OperandOwnership::PointerEscape:
+      return false;
+
+    case OperandOwnership::InstantaneousUse:
```

Comment:
Thanks for updating this to use OperandOwnership! This code was written before that was created IIRC.